_posts/2020-10-27-best-ec2-instance-for-rails-apps.md

```diff
@@ -0,0 +1,155 @@
+
+Because of the GIL, RRB runs with a balance of multiple processes and multiple threads-per-process. You can only run a single thread of Ruby code at once in a Ruby process, though multiple threads can be waiting on I/O or (usually) running code from a C-based native extension. Usually in Rails the balance comes to around 5 threads per process. I use [6 as a very minor improvement over 5 in RRB's specific case](https://rubykaigi.org/2017/presentations/codefolio.html). Ten Rails processes basically maxes out an EC2 2xlarge instances vCPUs and RAM for the amount of I/O waiting that Discourse is doing. So: ten processes, six threads/process.
+
+RRB is running code that is quite old &mdash; I designed it to benchmark the Ruby 3x3 transition, so compatibility with old code is a priority. On Christmas Day of 2020 Ruby 3 will be released and I can look at upgrading to recent versions of everything. In the mean time I'll keep using these old versions... And not timing EC2's [m6g ARM instances](https://aws.amazon.com/ec2/instance-types/m6/). I can build old Ruby on them, but other old code like ancient NodeJS and libv8 was really hard to get running.
+
+If you want to try m6g, you'll need to put some extra engineering time into it. You will likely have trouble getting and keeping it running since Intel processors are the default for everything. But if you have server bills large enough to make that a good idea, feel free to time it for yourself. My benchmark does ***not*** make that easy.
+
+So: special-case Rails apps may be a good fit for t-series, c-series or m6g-series instances. I'm not recommending them in general for everybody, and I'm not speed-testing them since they're usually the wrong answer for your Rails app.
+
+What does that leave? **Primarily it leaves m4 and m5 instances**. m4 is the older machine and architecture. They're similar in price.
+
+## Spot vs On-Demand Instances
+
+EC2 has [standard-price](https://aws.amazon.com/ec2/pricing/on-demand/) On-Demand Instances. You might or might not be able to get [Spot Instances](https://aws.amazon.com/ec2/spot/pricing/) at cheaper prices, randomly. Think of it as last-minute airline tickets. Sometimes you can get extra unused capacity cheap and sometimes you can't.
+
+So it's useful to know ***how much*** faster the new infrastructure (m5) is and then you can bid on it proportionately at appropriate prices.
+
+Amazon would like you to migrate off of m4, but there might be m4 spot instances available cheaply, depending. How does the price/performance compare? When would that be a good idea?
```

Comment:
> Amazon would like you to migrate off of m4, but there might be m4 spot instances available cheaply, depending

Is the "depending" a typo there? It looks out of place. 😄

Reply:
In effect I mean "sometimes" or "depending on other factors." It's an idiom, but I think a common one? Not opposed to changing it.

Reply:
I made it more explicit. Fixed.

Reply:
Awesome, thanks! 👍